### PR TITLE
Clean up database generators implementation

### PR DESCRIPTION
**Pull request by andrewn617**
### Motivation / Background

A lot of complexity has been added to the app generator and change generator for devcontainers, specifically around database setup. My goal here is to simplify this.

### Detail

First, I remove support for al the java and proprietary databases from these generators. Going forward, these generators will just support our core databases: `sqlite3`, `mysql`, `postgresql` and `trilogy`.

Then, I extracted all the database info into an object - gems, docker dependencies, and devcontainer setup. Then we just have to do `Database.build(options[:database])` to get all the needed database info for a generator. This reduces the huge amount of case statements littering the code.

A good side effect, is now the `ChangeGenerator` can just create its own database object. Previously, the `ChangeGenerator` was using the `Devcontainer` mixin to update the devcontainer database config. But that is sketchy because the change generator does not know everything about the app, such as if the app uses redis. So it _could_ be used to generate incorrect devcontainer config if someone started using those methods. Now that logic is separate from the database logic and the change generator only knows how to modify the DB setup.

This also allowed me to get rid of the hardcoded list of all the db features and service names which we need to iterate over in the change generator to delete the old db config.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: railties/lib/rails/generators/devcontainer.rb [hunk 1]
```diff
@@ -84,128 +78,180 @@ def devcontainer_needs_redis?
...
-        def db_name_for_devcontainer(database = options[:database])
-          case database
-          when "mysql"          then "mysql"
-          when "trilogy"        then "mariadb"
-          when "postgresql"     then "postgres"
-          end
+        def devcontainer_db_service_yaml(**options)
+          return unless service = devcontainer_database.service
+
+          { devcontainer_database.name => service }.to_yaml(**options)[4..-1]
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/devcontainer.rb:88]: What does 4 and 1 means here?

File: railties/lib/rails/generators/rails/db/system/change/change_generator.rb [hunk 2]
```diff
@@ -113,19 +112,17 @@ def edit_compose_yaml
...
+                compose_config["volumes"]&.delete(database.volume)
+                compose_config["services"]["rails-app"]["depends_on"]&.delete(database.name)
               end
 
-              db_service = db_service_for_devcontainer
-
-              if db_service
-                compose_config["services"].merge!(db_service)
-                compose_config["volumes"] = { db_volume_name_for_devcontainer => nil }.merge(compose_config["volumes"] || {})
+              if devcontainer_database.service
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/rails/db/system/change/change_generator.rb:121]: Can we go a step further and also refactor all the other places that have different behavior depending on the databsae in the generator? There is no reason why this object should be devcontaienr specific. We need to find the gem for the database, the docker image for the database, etc.

File: railties/lib/rails/generators/devcontainer.rb (hunk 1 above)
Comment (andrewn617) [reply railties/lib/rails/generators/devcontainer.rb:88]: It's the leading `---` that separate the body of the doc from directives. I don't know to_yaml takes an option to remove it.

File: railties/lib/rails/generators/rails/db/system/change/change_generator.rb (hunk 2 above)
Comment (andrewn617) [reply railties/lib/rails/generators/rails/db/system/change/change_generator.rb:121]: Yeah I can do that. The only slightly annoying thing is including all the other db that aren't available for devcontainer. I will work on it when I find some time this week.

Comment (rafaelfranca) [reply railties/lib/rails/generators/rails/db/system/change/change_generator.rb:121]: We can drop those databases. We already decided we would do, but never had a reason to. Now we have that reason.

File: railties/lib/rails/generators/devcontainer.rb [hunk 3]
```diff
@@ -84,128 +78,180 @@ def devcontainer_needs_redis?
...
+            source: Rails::Generators::RAILS_DEV_PATH,
+            target: Rails::Generators::RAILS_DEV_PATH
+          }
         end
 
-        def db_package_for_dockerfile(database = options[:database])
-          case database
-          when "mysql"          then "default-libmysqlclient-dev"
-          when "postgresql"     then "libpq-dev"
+        class Database
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/devcontainer.rb:99]: Can we move this entire class to railties/lib/rails/generators/database.rb? There is a lot of duplication still on that file that we can remove now that we have a abstraction to the database on generators.

Comment (andrewn617) [reply railties/lib/rails/generators/devcontainer.rb:99]: Yes this is still WIP

Comment (andrewn617) [reply railties/lib/rails/generators/devcontainer.rb]: @rafaelfranca Alright ready for another review

**Comment by headius** [comment]
This PR appears to have made it impossible to generate new Rails apps on JRuby without hand editing the bundle. What workflow should JRuby users follow now when generating new apps? Is there some way we can customize the "new" command for JRuby?

Alternatively, what can we do to restore the very few lines of code necessary to use the JDBC equivalents of the core adapters? We have commercial users interested in generating new Rails apps with JRuby.

**Pull request merged**